plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java

```diff
@@ -168,54 +169,175 @@ protected void canceling() {
+      return StatusUtil.error(getClass(), Messages.getString("save.credential.failed"), ex);
     }
+    return Status.OK_STATUS;
   }
 
-  private static void saveCredential(Path destination, Credential credential) throws IOException {
-    String jsonCredential = new CredentialHelper().toJson(credential);
-    Files.write(destination, jsonCredential.getBytes(StandardCharsets.UTF_8));
+  private IStatus stageProject(Path credentialFile, IPath explodedWarDirectory, IPath stagingDirectory, IProgressMonitor monitor) {
+    SubMonitor progress = SubMonitor.convert(monitor, 100);
+    StagingExitListener stagingExitListener = new StagingExitListener();
+    CloudSdk cloudSdk = getCloudSdk(credentialFile, stagingExitListener);
+
+    try {
+      getJobManager().beginRule(project, progress);
+      new ExplodedWarPublisher().publish(project, explodedWarDirectory, progress.newChild(40));
+      new StandardProjectStaging().stage(explodedWarDirectory, stagingDirectory, cloudSdk, progress.newChild(60));
+      return stagingExitListener.getExitStatus();
+    } catch (CoreException | IllegalArgumentException | OperationCanceledException ex) {
+      return StatusUtil.error(getClass(), Messages.getString("deploy.job.staging.failed"), ex);
+    } finally {
+      getJobManager().endRule(project);
+    }
+  }
+
+  private IStatus deployProject(Path credentialFile, IPath stagingDirectory, IProgressMonitor monitor) {
+    DeployExitListener deployExitListener = new DeployExitListener();
+    CloudSdk cloudSdk = getCloudSdk(credentialFile, deployExitListener);
+    new AppEngineProjectDeployer().deploy(
+        stagingDirectory, cloudSdk, deployConfiguration, monitor);
+    return deployExitListener.getExitStatus();
   }
 
-  private CloudSdk getCloudSdk(Path credentialFile) {
+  private CloudSdk getCloudSdk(Path credentialFile, ProcessExitListener processExitListener) {
     CloudSdk cloudSdk = new CloudSdk.Builder()
                           .addStdOutLineListener(stdoutLineListener)
                           .addStdErrLineListener(stderrLineListener)
                           .addStdErrLineListener(errorCollectingLineListener)
                           .appCommandCredentialFile(credentialFile.toFile())
                           .startListener(new StoreProcessObjectListener())
-                          .exitListener(new RecordProcessError())
+                          .exitListener(processExitListener)
                           .appCommandMetricsEnvironment(CloudToolsInfo.METRICS_NAME)
                           .appCommandMetricsEnvironmentVersion(CloudToolsInfo.getToolsVersion())
+                          .appCommandOutputFormat("json")
                           .build();
     return cloudSdk;
   }
 
+  private IStatus openAppInBrowser() {
+    final String project = deployConfiguration.getProject();
+    String appLocation = null;
+    if (deployConfiguration.getPromote()) {
+      appLocation = "http://" + project + ".appspot.com";
```

Comment:
Is it possible to push this logic in the `AppEngineDeployOutput` class?  Ideally this URL would be returned as part of the deploy output (IMHO).  Perhaps we can figure out whether to use the default URL using the `traffic_split` field?